KEP 498: Synchronized Startup Support for JobSets (#499)
kubernetes-sigs/jobset: keps/498-GroupScheduling/README.md

  - [Implmentation](#implmentation)
    - [Example JobSet spec AFTER webhook injection:](#example-jobset-spec-after-webhook-injection)
  - [Test Plan](#test-plan)
      - [Prerequisite testing updates](#prerequisite-testing-updates)
    - [Unit Tests](#unit-tests)
    - [Integration tests](#integration-tests)
  - [Graduation Criteria](#graduation-criteria)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
<!-- /toc -->

## Motivation

Many users of managed K8s services on cloud providers make use of NAP (node auto provisioning) which creates node pools for pending/unschedulable pods, based on those pods requirements (i.e., CPU/memory requirements, GPU/TPU requirements, etc).

Since node pool provisioning takes a variable amount of time, users are running into issues where the first slice finishes provisioning and pods land there and begin running, but eventually timeout before the other slices all finish provisioning and pods land there and become ready.

For statically provisioned clusters, we have recommended users use Kueue to handle group scheduling of JobSets once sufficient resources are available.

> However, for dynmaically provisioned clusters, we have no support for group scheduling in JobSet yet.

Reviewer: So kueue will support dynamically provisioned clusters if there is support in it for jobset?

Author: I don't think Kueue will support dynamic provisioning, we need pods to be pending in order to trigger the dynamic provisioning for them, but Kueue keeps the Jobs in a suspended state (i.e., no pods created) until there is sufficient resources for them. 

I meant we have no group scheduling support for JobSets that are being run independently without Kueue. This specifically becomes an issue on clusters using dynamic node pool provisioning. I updated this section for clarity.

### Goals

> - Enhance the Failure Policy API with options allowing the user to define a group scheduling policy.

Author: Typo, fixed.

### Non-Goals

- Upstream changes that would improve/simplify group scheduling support in JobSet.

## Proposal

> ### User Stories (Optional)

Reviewer: For each of these stories, can you represent what the API will look like?

Author: Sure, added

#### Story 1: Group scheduling of all replicated jobs

As a user, in order to make efficient use of expensive accelerator (GPU/TPU) resources, I use node
auto-provisioning to provision infrastructure on an as-needed basis when a pending workload requires
it, then deprovision it once it's no longer in use. Since the completion time of these provisioning
operations is variable, I want to make sure JobSet pods which land on the earliest provisioned nodes
do not start executing the main workload container until all the infrastructure is provisioned and
all pods have started. Otherwise, the distributed initialization step will timeout in the earliest
pods while waiting for the remaining provisioning to complete.

#### Story 2: Group scheduling of a specific replicated job

As a user, I have a JobSet with 2 replicated jobs:
- `workers` which contains my primary batch workload, running on nodes with accelerator chips (GPUs)

> - `auxiliary` which contains auxiliary pods (proxy server, metrics service) running on CPU nodes.

Author: Fixed

I want my batch workload workers to be scheduled as a group, but the auxiliary pods can start up
individually at any time.

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

### Proposed Group Scheduling API

```go

// JobSetSpec defines the desired state of JobSet
type JobSetSpec struct {
        …

>         // GroupSchedulingConfig defines the desired group-scheduling behavior

Reviewer: ```suggestion
        // GroupSchedulingConfig defines the desired group-scheduling behavior
        // for the JobSet. If unspecified, the replicated jobs will be started without any waiting.
```

Something like that.

Author: Updated.

        // for the JobSet. If unspecified, pods will begin running as soon as
        // they are scheduled and ready.
        GroupSchedulingConfig *GroupSchedulingConfig `json:groupSchedulingConfig`
}

// GroupSchedulingConfig defines the desired group-scheduling behavior for the

> // JobSet. If defined, the main workload container in pods part of the

Reviewer: I don't understand this comment.

What is a main workload container?

Author: The container running the user workload (not the initContainer). 

Updated this to `user container` everywhere for clarity. Just trying to distinguish between the initContainer and user container.

    after a short interval.
 4. Once the expected number of pods for the JobSet have started (derived from the
    JobSet spec), the pod controller will update the ConfigMap with some arbitrary data
    so that it is no longer empty. This will trigger Kubelet to populate
    the initContainer mount point. 
 5. Now that the mount point directory path exists in the initContainer, it
    will exit and allow the main workload container to start.

#### Example JobSet spec AFTER webhook injection:

```yaml
apiVersion: jobset.x-k8s.io/v1alpha2
kind: JobSet
metadata:
  name: my-jobset
  namespace: default
spec:
  # main containers in pods part of `workers` replicatedJob will be blocked
  # from execution until all pods have started.
  groupSchedulingConfig:
    targetReplicatedJobs:
    - workers
  replicatedJobs:
    - name: workers
      replicas: 3
      template:
        spec:
          backoffLimit: 0
          completions: 100
          parallelism: 100
          template:
            spec:
              initContainers:
                # Check if the started file is present before exiting and starting the main container.

>                 - name: group-scheduling-init

Reviewer: We probably should mention what names will be forbidden for the "main pod"?

I'm not sure if the name is taking by the main container, will the init container take perference over the main container? Or will they have conflicts.

Author: Good point, I will add something to "constraints" section about this. We can validate this in the JobSet webhook to ensure there's no conflicts.

>                   image: bash

Reviewer: Will we provide configuration on the init container that gets injected?

Reviewer: Generally I wonder about the image, registry settings, image pull secrets.

Author: I was thinking about either just using `bash:latest` (or bash at a specific tag we've tested at). 

The command executed in the initContainer will be injected by the JobSet webhook, see the example of the JobSet spec after injection in the implementation section.

Reviewer: Yea, I'm concerned that some users may require that all containers have image pull secrets. Or they want to use a special registry for their containers. Maybe they want to avoid pulling images from dockerhub or want to use quay as their registry.

Reviewer: Maybe this is YANGI..

Reviewer: For example, in openshift, we also have a lot of cases where people use security context and volumes may have selinux label issues.

The testing plan will focus on:
- Unit and integration tests to validate the webhook injection works as intended
- E2E tests, since we need pods to be created to injectthe initContainer and ConfigMap volume mount.
- Scale tests, performance benchmarking

##### Prerequisite testing updates

<!--
Based on reviewers feedback describe what additional tests need to be added prior
implementing this enhancement to ensure the enhancements have also solid foundations.
-->

#### Unit Tests

<!--
In principle every added code should have complete unit test coverage, so providing
the exact set of tests will not bring additional value.
However, if complete unit test coverage is not possible, explain the reason of it
together with explanation why this is acceptable.
-->

<!--
Additionally, try to enumerate the core package you will be touching
to implement this enhancement and provide the current unit coverage for those
in the form of:
- <package>: <date> - <current test coverage>

This can inform certain test coverage improvements that we want to do before
extending the production code to implement this enhancement.
-->

- `controllers`: `04/05/2024` - `25.4%`

#### Integration tests

> -

Reviewer: We should probably mention some test cases here.

Author: Added

it, then deprovision it once it's no longer in use. Since the completion time of these provisioning
operations is variable, I want to make sure JobSet pods which land on the earliest provisioned nodes
do not start executing the main container until all the infrastructure is provisioned and
all pods have started. Otherwise, the distributed initialization step will timeout in the earliest
pods while waiting for the remaining provisioning to complete.

### Example JobSet spec using group scheduling for all replicated jobs:

```yaml
apiVersion: jobset.x-k8s.io/v1alpha2
kind: JobSet
metadata:
  name: my-jobset
  namespace: default
spec:
  # main containers in pods part of all replicatedJob will be blocked
  # from execution until all pods have started, or timeout after 5min.
  groupSchedulingConfig:
    timeoutAfterSeconds: 300
  replicatedJobs:
    - name: workers
      replicas: 3
      template:
        spec:
          backoffLimit: 0
          completions: 100
          parallelism: 100
          template:
            spec:
              containers:
                - name: pytorch
                  image: pytorch/pytorch:latest
                  command:

>                     - bash

Reviewer: its here so I'll comment on it..

I don't think bash will run a python script. :)

Author: Woops! Fixed

            spec:
              initContainers:
                # Check if the started file is present before exiting and starting the main container.
                - name: group-scheduling-init
                  image: bash
                  command: # TODO: add support for configurable timeout parameter
                    - bash
                    - -c
                    - while true; do if [ -f /mnt/group-scheduling/started ]; then break; fi; sleep 1; done
                  volumeMounts:
                    - mountPath: /mnt/group-scheduling
                      name: group-scheduling
              containers:
                - name: pytorch
                  image: pytorch/pytorch:latest
                  command:
                    - bash
                    - -c
                    - train.py
              volumes:
                - name: group-scheduling
                  configMap:
                    # Format: {jobset-name}-started-threshold
                    name: my-jobset-group-scheduling
                    optional: true
```

### Test Plan

The testing plan will focus on:
- Unit and integration tests to validate the webhook injection works as intended

> - E2E tests, since we need pods to be created to injectthe initContainer and ConfigMap volume mount.

Reviewer: ```suggestion
- E2E tests, since we need pods to be created to inject the initContainer and ConfigMap volume mount.
```

Author: Fixed

  - [Example JobSet spec using group scheduling for all replicated jobs:](#example-jobset-spec-using-group-scheduling-for-all-replicated-jobs-1)
  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
  - [Risks and Mitigations](#risks-and-mitigations)
- [Design Details](#design-details)
  - [Proposed Group Scheduling API](#proposed-group-scheduling-api)
  - [Constraints](#constraints)
  - [Implmentation](#implmentation)
    - [Example JobSet spec AFTER webhook injection:](#example-jobset-spec-after-webhook-injection)
  - [Test Plan](#test-plan)
      - [Prerequisite testing updates](#prerequisite-testing-updates)
    - [Unit Tests](#unit-tests)
    - [Integration tests](#integration-tests)
  - [Graduation Criteria](#graduation-criteria)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
<!-- /toc -->

## Motivation

Many users of managed K8s services on cloud providers make use of NAP (node auto provisioning) which creates node pools for pending/unschedulable pods, based on those pods requirements (i.e., CPU/memory requirements, GPU/TPU requirements, etc).

> Since node pool provisioning takes a variable amount of time, users are running into issues where the first slice finishes provisioning and pods land there and begin running, but eventually timeout before the other slices all finish provisioning and pods land there and become ready.

Reviewer 2: We discussed other motivations other than provisioning, could you please add them?

> For statically provisioned clusters, we have recommended users use Kueue to handle group scheduling of JobSets once sufficient resources are available.

Reviewer: Maybe @tenzen-y, @alculquicondor or @mimowo can comment on Kueue and dynamic clusters?

I know that wg-batch did some work in autoscaling and I'm curious how this feature is related.

Reviewer: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/proposals/provisioning-request.md I think was the proposal.

Author: I'm curious as well. For additional context for the Google folks, we are talking about NAP (not CA) here.

> However, JobSets running independently (without Kueue) on dynmaically provisioned clusters, we need to support group

Reviewer: ```suggestion
However, JobSets running independently (without Kueue) on dynamically provisioned clusters, we need to support group
```

> scheduling semantics in order to avoid these timeout issues.

Reviewer 2: ```suggestion
startup semantics in order to avoid these timeout issues.
```

### Goals

- Enhance the JobSet API with options allowing the user to define a group scheduling policy.

### Non-Goals

- Upstream changes that would improve/simplify group scheduling support in JobSet.

## Proposal

### User Stories (Optional)

> #### Story 1: Group scheduling of all replicated jobs

Reviewer 2: ```suggestion
#### Story 1: Group startup of all replicated jobs
```

As a user, in order to make efficient use of expensive accelerator (GPU/TPU) resources, I use node

> auto-provisioning to provision infrastructure on an as-needed basis when a pending workload requires

Reviewer 2: the node auto-provisioner is a GKE specific. However, similar problem arise in any autoscaled environment.

operations is variable, I want to make sure JobSet pods which land on the earliest provisioned nodes
do not start executing the main container until all the infrastructure is provisioned and
all pods have started. Otherwise, the distributed initialization step will timeout in the earliest
pods while waiting for the remaining provisioning to complete.

### Example JobSet spec using group scheduling for all replicated jobs:

```yaml
apiVersion: jobset.x-k8s.io/v1alpha2
kind: JobSet
metadata:
  name: my-jobset
  namespace: default
spec:
  # main containers in pods part of all replicatedJob will be blocked
  # from execution until all pods have started, or timeout after 5min.
  groupSchedulingConfig:
    timeoutAfterSeconds: 300
  replicatedJobs:
    - name: workers
      replicas: 3
      template:
        spec:
          backoffLimit: 0
          completions: 100
          parallelism: 100
          template:
            spec:
              containers:
                - name: pytorch
                  image: pytorch/pytorch:latest
                  command:
                    - python3
                    - train.py
```

> #### Story 2: Group scheduling of a specific replicated job

Reviewer 2: ```suggestion
#### Story 2: Group startup of a specific replicated job
```

As a user, I have a JobSet with 2 replicated jobs:
- `workers` which contains my primary batch workload, running on nodes with accelerator chips (GPUs)
- `auxiliary` which contains auxiliary workloads (proxy server, metrics service) running on CPU nodes.

> I want my batch workload workers to be scheduled as a group, but the auxiliary pods can start up

Reviewer 2: ```suggestion
I want my batch workload workers to start as a group, but the auxiliary pods can start up
```

Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

### Proposed Group Scheduling API

```go

// JobSetSpec defines the desired state of JobSet
type JobSetSpec struct {
        …
        // GroupSchedulingConfig defines the desired group-scheduling behavior
        // for the JobSet. If unspecified, every pod part of this JobSet will
        // start as soon as possible, without waiting for the others.
        GroupSchedulingConfig *GroupSchedulingConfig `json:groupSchedulingConfig`
}

// GroupSchedulingConfig defines the desired group-scheduling behavior for the
// JobSet. If defined, the main container in pods part of the
// TargetReplicatedJobs will be blocked from executing until all pods in these
// jobs have started.
type GroupSchedulingConfig struct {
        // Timeout defines the period after which the injected initContainer
        // (which blocks execution of the main container until all pods are started)
        // will timeout and exit with an error if not all pods have started yet.

>         TimeoutAfterSeconds *int32 `json:"timeoutAfterSeconds"

Reviewer: Reading 5 minutes to mean 300 s I wonder if we should consider durations as an API?

I guess k8s has a lot of apis in seconds so maybe its not a big deal.

        // TargetReplicatedJobs are the names of the replicated jobs which will
        // be subject to group-scheduling.
        // A null or empty list will apply to all replicatedJobs.
        // +optional
        // +listType=atomic
        TargetReplicatedJobs []string `json:"targetReplicatedJobs,omitempty"`
}
```

### Constraints

> - In Order Startup Policy is incompatible with a group scheduling config that spans

Reviewer: I think we should comment on when someone should use startup policy or this feature.
